components/style/values/specified/length.rs

```diff
@@ -4,44 +4,121 @@
 
 use app_units::Au;
 use cssparser::{Parser, Token};
-use euclid::size::Size2D;
 use font_metrics::FontMetrics;
 use parser::{Parse, ParserContext};
 use std::ascii::AsciiExt;
 use std::cmp;
 use std::fmt;
+use std::mem;
 use std::ops::Mul;
 use style_traits::ToCss;
 use style_traits::values::specified::AllowedNumericType;
 use super::{Angle, Number, SimplifiedValueNode, SimplifiedSumNode, Time};
-use values::{Auto, CSSFloat, Either, FONT_MEDIUM_PX, HasViewportPercentage, None_, Normal};
+use values::{Auto, Either, None_, Normal};
+use values::{CSSFloat, FONT_MEDIUM_PX, HasViewportPercentage, NoViewportPercentage};
 use values::computed::Context;
 
 pub use super::image::{AngleOrCorner, ColorStop, EndingShape as GradientEndingShape, Gradient};
 pub use super::image::{GradientKind, HorizontalDirection, Image, LengthOrKeyword, LengthOrPercentageOrKeyword};
 pub use super::image::{SizeKeyword, VerticalDirection};
 
+const NUM_FONT_UNITS: usize = 4;
+const NUM_VIEWPORT_UNITS: usize = 4;
+
+#[repr(u8)]
+#[derive(Clone, PartialEq, Copy, Debug)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub enum FontUnit {
```

Comment:
I've got multiple choices here:

 - `define_css_keyword_enum` and `define_css_numbered_enum` are both useful here, but it'll also impl a `parse` method (the generic one that uses CSS parser). It won't be useful for us, because we need something like the `parse_dimension` below.
 - Or, we could have a new macro for this, but I wonder how good/useful that's going to be.